neos/neos-ui
# TASK: Add another way to run tests locally with a system under test in docker

**What I did**
I made it possible to start a neos test instance as _system under test_ (SUT) via docker. You are able to run testcafe tests on your machine to faster implement tests and features.

**How I did it**
Add a shell script to start the SUT.
Add a make command to do it fast.
Add some configs to make it possible.

We Mount the NeosUi root into the SUT.
Then we symlink the `[...]/Packages/Application/Neos.Ui` to this mounted volume.
We do the same for `[...]/Packages/Application/Neos.TestNodeTypes` -> `[...]/Tests/IntegrationTests/SharedNodeTypesPackage`.

This way we can build tests, features and bugfixes without the need to restart the SUT (in 99% of cases).
This speeds up the feedback loop by at least an order of magnitude. It's insane.

We don't even need to have a Neos Instance to develop the UI! I think this could be huge for new contributors.
Just start the SUT and you have a running Neos 🚀 

If you have to nuke it, just re-run the make command.

**Why I did it**
It's extremely slow to run tests on my local machine via docker because it completely wipes and reinstalls neos and all the things everytime I want to run 1 or 2 tests.
I decoupled the SUT and "running the tests". This makes writing tests and features so much faster for me.

**How to verify it**
- `make e2e-start-system-under-test`
- `yarn run testcafe chrome Tests/IntegrationTests/Fixtures -T 'sidebars'` - tests should not fail
- change something that breaks that feature (I just broke the LeftSideBar reducer)
- `make build-e2e-testing` - rebuild Ui
- `yarn run testcafe chrome Tests/IntegrationTests/Fixtures -T 'sidebars'` - tests should fail
